### ISSUE TITLE
Main form must validate all sub form data before submission

### ISSUE THREAD

**Issue by HauklandJ** [Altinn/app-frontend-react#2148]
## Update
The below description and diagram is possibly outdated. Read comments and investigate as required before tackling this task.

## Blocked by
- https://github.com/Altinn/app-frontend-react/issues/1484

## Description
Part of the [Underskjema Epic](https://github.com/Altinn/app-frontend-react/issues/2137)

A user should NOT be able to submit a form if one or more sub forms are invalid. Two possible solutions:

1. Validate form and sub forms on `process next`
2. Validate sub forms on `sub form close` and check validation state of sub forms + validate form on `process next`

![validation2 drawio](https://github.com/Altinn/app-frontend-react/assets/42615991/593c14dd-1bb6-4eff-a6e7-f821a96e2596)

Option 1 is straight-forward, but may lead to performance issues in the case of many sub forms.

Option 2 is more complex, but would be a better user experience in the case of many sub forms. It would include computing and persisting a hash of the validated sub form on `sub form close` and then computing hashes of the sub forms on `form submit` and verifying against the persisted sub form hashes

**Comment by bjosttveit** [comment]
Just throwing some thoughts on this:

The way validation in frontend works currently, is that we validate everything initially, and then incrementally as data in the form changes. We dont actually validate explicitly before submitting, we just check the current validation state which should be up to date at that point. However, the backend will validate everything when we call process next. I am assuming this will be the case for all sub forms in the task as well?

I am not exactly sure what the best model for sub forms would be. What if you refresh the page before submitting? Then you would not have any state from having been inside of the sub form and closing it, so there probably needs to be some mechanism for validating all sub forms from the "outside"? But whether it should be done on submit or in a similar way to the existing validations I do not know.

**Comment by HauklandJ** [comment]
> Just throwing out some thoughts on this:

> The way validation in frontend works currently, is that we validate everything initially, and then incrementally as data in the form changes. We dont actually validate explicitly before submitting, we just check the current validation state which should be up to date at that point. However, the backend will validate everything when we call process next. I am assuming this will be the case for all sub forms in the task as well?
> 
We toying with the possibility of validating via backend when a sub form is closed with the "Add/Legg til" button. This would include a hash of the sub form content. Whether this is a performance boost that is needed is something we value input on, given the extra complexity introduced.
> I am not exactly sure what the best model for sub forms would be. What if you refresh the page before submitting? Then you would not have any state from having been inside of the sub form and closing it, so there probably needs to be some mechanism for validating all sub forms from the "outside"? But whether it should be done on submit or in a similar way to the existing validations I do not know.

[There is another issue pertaining to this](https://github.com/Altinn/app-frontend-react/issues/2153). Would like your input there!

**Comment by danielskovli** [comment]
For what its worth, I'm liking the `option 2` approach. The sub form table component would hold a reference to all sub forms and their validation status (flag + digest). Upon submission the backend would confirm the digest if the status is set to `valid`, or fall back to a complete validation routine if the status missing/invalid or the digest is outdated.

This way, in a linear flow, the user would only get hit with validation every time they finish editing a sub form, then again for the main for upon submission.

Likely edge-case: If a form pre-fills a large amount of sub form data, the user would have to enter each of the sub forms to validate the data. But this is perhaps the desired outcome for most pre-fill scenarios? We could also allow the pre-fill service to also set a validation status, if they believe all required data is included without the requirement of any user input. 🤷

**Comment by bjosttveit** [comment]
Are you saying we should prevent the user from submitting until they have entered each sub form so they will get validated? What if you refresh the page after filling everything out, or maybe the user opens the instance on a different machine. It seems like a bad experience having to reopen every sub form then.

Additionally, validations from backend are usually not as good as the ones the frontend generates, for example, backend data model validations are not translated, so they should really only be shown as a last resort.

**Comment by danielskovli** [comment]
No I wasn't trying to say that, but I see now that perhaps I have a slightly naive understanding of how validation works. In my scenario above, the server would hold the validation status as metadata. If front-end loses track of the status, it just needs to ask again.

As for front-end vs back-end valiation, that's an interesting point. I have no good response to that.

**Comment by bjosttveit** [comment]
Ait, I thought that was what you meant in the edge case scenario

**Comment by danielskovli** [comment]
Just bumping this after a brief discussion with @olemartinorg and @HauklandJ earlier today.

There are no new insights to add, just  fact that this is perhaps the trickiest remaining topic standing between us and an initial release of the whole sub-form concept.

Did you want to be more involved with this topic @bjosttveit?

**Comment by bjosttveit** [comment]
> Did you want to be more involved with this topic @bjosttveit?

Sure, I am not sure I am up to speed now though, or, at least I haven't thought about it for a while 😅

**Comment by olemartinorg** [comment]
I had a look at the sub-form test app, and I've been trying out the concepts. The solution looks great! :raised_hands: But yep, the way validations work right now is somewhat problematic.

The way I see it, we have some requirements (not all of these may block our ability to ship the first version of this feature, though):
1. While filling out a sub-form, one should be able to control when validations in that sub-form show up for different components (solved)
2. It should be possible to close the sub-form and an app developer should be able to configure if closing the sub-form should be blocked in case there are validation errors (solved)
3. When viewing a list of sub-forms in the main form, we should indicate whether or not each sub-form properly validates or not (not solved)
4. When submitting the main form, if there are validation errors in a sub-form these should show up (kinda solved, but the error messages are not the ones we'd normally use in app-frontend)
5. When submitting the main form, if there are validation errors in a sub-form you should be able to click on that error message and be transported to the correct field inside the sub-form, or at least the sub-form itself (not solved)
6. It should be possible to validate across the main form and sub-forms, i.e. validate that the sum of all numbers in all sub-forms equals the expected sum entered in the main form (not solved? I have no idea)

So, right now nodes are not created for sub-forms when in the main form. I think this is the correct approach, but validations are often tied to nodes (which makes it possible to link directly to the node in the `ErrorReport` component, for example). Not having any nodes available for a sub-form in the main-form context might prevent us from easily implementing that, but at the same time we'd have to think about how error messages are presented when forms grow as large as they can with sub-forms. For example, do we really have to deep-link to a node in a sub-form, or could we show a 'one or more errors in this sub-form, click here to open the sub-form and see the full list'? Showing very specific validations in the main form without their context is confusing - 'name is a required field' x100 is not as helpful as we like to think.

**Comment by bjosttveit** [comment]
>     1. While filling out a sub-form, one should be able to control when validations in that sub-form show up for different components (solved)
> 
>     2. It should be possible to close the sub-form and an app developer should be able to configure if closing the sub-form should be blocked in case there are validation errors (solved)
> 
>     3. When viewing a list of sub-forms in the main form, we should indicate whether or not each sub-form properly validates or not (not solved)
> 
>     4. When submitting the main form, if there are validation errors in a sub-form these should show up (kinda solved, but the error messages are not the ones we'd normally use in app-frontend)
> 
>     5. When submitting the main form, if there are validation errors in a sub-form you should be able to click on that error message and be transported to the correct field inside the sub-form, or at least the sub-form itself (not solved)
> 
>     6. It should be possible to validate across the main form and sub-forms, i.e. validate that the sum of all numbers in all sub-forms equals the expected sum entered in the main form (not solved? I have no idea)

Everything except for 6 is solved after #2479 . As for 4 and 5, we don't show the individual validation errors for subforms when inside of the main form, only that there are one or more errors with a set of subforms. Therefore it only transports you to the sub-form component, not to the field(s) inside of it/them.
